cmsl1t/collections/vectorized.py

```diff
@@ -0,0 +1,167 @@
+import awkward
+from collections import defaultdict
+import logging
+import numpy as np
+import random
+from rootpy.plotting import Hist
+
+from . import BaseHistCollection
+from ..utils.iterators import pairwise
+from .. import PY3
+from ..io import to_root
+
+logger = logging.getLogger(__name__)
+
+
+def extend(arr1, starts, stops):
+    repeat = stops - starts
+    return np.repeat(arr1, repeat, axis=0)
+
+
+def split_input(inner_indices, x, w):
+    content = x
+    if hasattr(x, 'starts'):
+        inner_indices = extend(inner_indices, x.starts, x.stops)
+        content = x.content
+
+    for u in np.unique(inner_indices):
+        mask = inner_indices == u
+        if not isinstance(mask, (list, np.ndarray)):
+            mask = np.array([mask])
+        yield u, content[mask], w[mask]
+
+
+class VectorizedHistCollection(BaseHistCollection):
+
+    def __init__(self, innerBins, innerLabel='inner', **kwargs):
+        # if we want to generalize to N dim, innerBins needs to be an array of innerBins
+        # TODO: last dimension should probably be a normal dictionary
+        dimensions = kwargs.pop('dimensions', 2)
+        self._name = kwargs.pop('name', str(hex(random.getrandbits(128)))[2:10])
+        self._execute_before_write = kwargs.pop('execute_before_write', [])
+        if PY3:
+            super(VectorizedHistCollection, self).__init__(dimensions)
+        else:
+            BaseHistCollection.__init__(self, dimensions)
+
+        self._innerBins = innerBins
+        self._innerLabel = innerLabel
+        self._innerHist = Hist(100, 0, 100, name=innerLabel + '_' + self._name)
+
+    def __getitem__(self, key):
+        if not isinstance(key, (list, np.ndarray, np.generic)):
+            key = np.array([key])
+        real_keys = self._get_inner_indices(key)
+        return VectorizedBinProxy(self, real_keys)
+        # return [defaultdict.__getitem__(self, k) for k in real_keys.tolist()]
+
+    def _get_inner_indices(self, values):
+        '''
+            Returns the pileup bin corresponding to the provided pileup value.
+             - bin 0 is underflow
+             - bin len(innerBins) is overflow
+
+            :Example:
+                >>> hists = VectorizedHistCollection(innerBins=[0,10,15,20,30,999])
+                >>> hists._get_inner_indices([1, 11, 1111]) # returns [1, 2, 6]
+        '''
+        return np.digitize(values, self._innerBins)
+
+    def add(self, name, bins, hist_type=Hist, **kwargs):
+        title = kwargs.pop('title', name)
+        bins = np.asarray(bins)
+        if bins.size == 0:
+            logger.error(
+                'No bins specified for histogram {0}'.format(name))
+
+        if name in defaultdict.__getitem__(self, 1):
```

Comment:
This baffled me for a while, but then I realised that `defaultdict` was a super-class of this one.  But it's twice removed from here, and so this seems like quite a brittle construction.  I think in the current form of the code:
```python
if name in super(VectorizedHistCollection, self)[1]:
```
should do the same thing but be more robust to changes in the super-classes.

Reply:
That's interesting, I will give that a go

Reply:
```
TypeError: 'super' object is not subscriptable
```
but
```python
super(VectorizedHistCollection, self).__getitem__(1)
```
works